File: solution/138/solution.py
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution(object):
    def copyRandomList(self, head):
        oldmap = {}
        newlist = []
        i = 0
        ptr = head
        while ptr:
            oldmap[ptr] = i
            newlist.append(RandomListNode(ptr.label))
            i += 1
            ptr = ptr.next
        newlist.append(None)
        i = 0
        ptr = head
        while ptr:
            newlist[i].next = newlist[i+1]
            if ptr.random:
                index = oldmap[ptr.random]
                newlist[i].random = newlist[index]
            i += 1
            ptr = ptr.next
        return newlist[0]

File: solution/138/solution.cpp
#include <stdio.h>
#include <map>
#include <vector>

struct RandomListNode{
    int label;
    ListNode * next, random;
    RandomListNode(int x):label(x), next(NULL), random(NULL){}
};


class Solution{
public:
    RandomListNode* copyRandomList(RandomListNode* head){
        std::map<RandomListNode*, int> oldmap;
        std::vector<RandomListNode*> newmap;
        RandomListNode *ptr = head;
        int i = 0;
        while(ptr){
            oldmap[ptr] = i;
            newmap.push_back(new RandomListNode(ptr->label));
            ptr = ptr->next;
            i ++;
        }
        newmap.push_back(NULL);
        ptr = head;
        i = 0;
        while(ptr){
            newmap[i]->next = newmap[i+1];
            if(ptr->random){
                int index = oldmap[ptr->random];
                newmap[i]->random = newmap[index];
            }
            ptr = ptr->next;
            i++;
        }
        return newmap[0];
    }
}